lib/remote_node.go

```diff
@@ -0,0 +1,468 @@
+	rn.connectionStatus = RemoteNodeConnectionStatus_Attempted
+}
+
+// ConnectInboundPeer connects a peer once a successful inbound connection has been established.
+func (rn *RemoteNode) ConnectInboundPeer(conn net.Conn, na *wire.NetAddress) error {
+	peer := rn.cmgr.ConnectPeer(conn, na, RemoteNodeIdNoAttempt, false, false)
+	if peer == nil {
+		return errors.Errorf("ConnectInboundPeer: Problem connecting peer (%s)", conn.RemoteAddr().String())
+	}
+	rn.SetConnectedPeer(peer)
+	return nil
+}
+
+// ConnectOutboundPeer connects a peer once a successful outbound connection has been established.
+func (rn *RemoteNode) ConnectOutboundPeer(conn net.Conn, na *wire.NetAddress, attemptId uint64, isOutbound bool, isPersistent bool) error {
+	peer := rn.cmgr.ConnectPeer(conn, na, attemptId, isOutbound, isPersistent)
+	if peer == nil {
+		return errors.Errorf("ConnectInboundPeer: Problem connecting peer (%s)", conn.RemoteAddr().String())
+	}
+	rn.SetConnectedPeer(peer)
+	return nil
+}
+
+func (rn *RemoteNode) Disconnect() {
+	peerId, attemptId := rn.id.GetIds()
+	switch rn.connectionStatus {
+	case RemoteNodeConnectionStatus_Attempted:
+		rn.cmgr.CloseAttemptedConnection(attemptId)
+	case RemoteNodeConnectionStatus_Connected:
+		rn.cmgr.CloseConnection(peerId)
+	}
+	rn.connectionStatus = RemoteNodeConnectionStatus_Terminated
+}
+
+func (rn *RemoteNode) SendMessage(desoMsg DeSoMessage) {
+	if rn.connectionStatus != RemoteNodeConnectionStatus_Connected {
+		return
+	}
+
+	if err := rn.cmgr.SendMessage(desoMsg, rn.GetPeerId()); err != nil {
+		glog.Errorf("sendMessage: Problem sending message to peer (id= %d): %v", rn.peer.ID, err)
+	}
+}
+
+func (rn *RemoteNode) GetHandshakeMetadata() *HandshakeMetadata {
+	if rn.handshakeMetadata == nil {
+		rn.handshakeMetadata = &HandshakeMetadata{}
+	}
+	return rn.handshakeMetadata
+}
+
+func (rn *RemoteNode) InitiateHandshake(nonce uint64) error {
+	if rn.GetPeer() == nil {
+		return errors.Errorf("Remote node has no peer")
+	}
+
+	if rn.GetPeer().IsOutbound() {
+		vMeta := rn.GetHandshakeMetadata()
+		versionTimeExpected := time.Now().Add(rn.params.VersionNegotiationTimeout)
+		vMeta.versionTimeExpected = &versionTimeExpected
+		rn.sendVersionMessage(nonce)
+	}
+	return nil
+}
+
+// sendVersionMessage generates and sends a version message to a RemoteNode peer. The message will contain the nonce
+// that is passed in as an argument.
+func (rn *RemoteNode) sendVersionMessage(nonce uint64) {
+	verMsg := rn.newVersionMessage(nonce)
+
+	// Record the nonce of this version message before we send it so we can
+	// detect self connections and so we can validate that the peer actually
+	// controls the IP she's supposedly communicating to us from.
+	vMeta := rn.GetHandshakeMetadata()
+	vMeta.versionNoncesSent = nonce
+	//uint64(RandInt64(math.MaxInt64))
+
+	if err := rn.cmgr.SendMessage(verMsg, rn.peer.ID); err != nil {
+		glog.Errorf("sendVersionMessage: Problem sending version message to peer (id= %d): %v", rn.peer.ID, err)
```

Comment:
Should we return this error instead of glogging it? Is there ever a scenario in which the caller would want to handle this error?